src/meta/src/manager/catalog/mod.rs

```diff
@@ -836,66 +838,296 @@ where
 
-                        for user in users_need_update {
-                            self.notify_frontend(Operation::Update, Info::User(user))
-                                .await;
-                        }
+                    // index table and index.
+                    user_core.decrease_ref_count(index.owner, 2);
 
-                        self.notify_frontend(Operation::Delete, Info::Table(table))
+                    for user in users_need_update {
+                        self.notify_frontend(Operation::Update, Info::User(user))
                             .await;
+                    }
 
-                        for dependent_relation_id in dependent_relations {
-                            database_core.decrease_ref_count(dependent_relation_id);
-                        }
-
-                        let version = self
-                            .notify_frontend(Operation::Delete, Info::Index(index))
-                            .await;
+                    self.notify_frontend(Operation::Delete, Info::Table(table))
+                        .await;
 
-                        Ok(version)
+                    for dependent_relation_id in dependent_relations {
+                        database_core.decrease_ref_count(dependent_relation_id);
                     }
+
+                    let version = self
+                        .notify_frontend(Operation::Delete, Info::Index(index))
+                        .await;
+
+                    Ok(version)
                 }
-            } else {
-                bail!("index table doesn't exist",)
             }
         } else {
-            bail!("index doesn't exist",)
+            bail!("index table doesn't exist",)
         }
     }
 
+    pub async fn alter_table_name(
+        &self,
+        table_id: TableId,
+        table_name: &str,
+    ) -> MetaResult<NotificationVersion> {
+        let core = &mut *self.core.lock().await;
+        let database_core = &mut core.database;
+        database_core.ensure_table_id(table_id)?;
+
+        // 1. validate new table name.
+        let mut table = database_core.tables.get(&table_id).unwrap().clone();
+        let old_name = table.name.clone();
+        database_core.check_relation_name_duplicated(&(
+            table.database_id,
+            table.schema_id,
+            table_name.to_string(),
+        ))?;
+
+        let source = table.optional_associated_source_id.as_ref().map(
+            |OptionalAssociatedSourceId::AssociatedSourceId(id)| {
+                let mut source = database_core.sources.get(id).unwrap().clone();
+                source.name = table_name.to_string();
+                source
+            },
+        );
+
+        // 2. rename table and its definition.
+        table.name = table_name.to_string();
+        table.definition = alter_relation_rename(&table.definition, table_name);
+
+        // 3. update all relations that depend on this table, note that indexes are not included.
```

Comment:
Why do we ignore indexes here? 👀

Reply:
Because we don't store index definition in `Index` catalog. If it is to be supported subsequently it can be deduced from its catalog.